python/ray/_private/runtime_env/pip.py

```diff
@@ -428,16 +433,25 @@ async def create(
         target_dir = self._get_path_from_hash(hash)
 
         async def _create_for_hash():
-            await PipProcessor(target_dir, runtime_env, logger)
+            await PipProcessor(
+                target_dir,
+                runtime_env,
+                logger,
+            )
 
             loop = get_running_loop()
             return await loop.run_in_executor(
                 None, get_directory_size_bytes, target_dir
             )
 
-        self._creating_task[hash] = task = create_task(_create_for_hash())
-        task.add_done_callback(lambda _: self._creating_task.pop(hash, None))
-        return await task
+        if uri not in self._create_locks:
+            # async lock to prevent the same virtualenv being concurrently installed
+            self._create_locks[uri] = asyncio.Lock()
+
+        async with self._create_locks[uri]:
+            self._creating_task[hash] = task = create_task(_create_for_hash())
```

Comment:
Why don't you set a flag here and test before starting? That way subsequent creates will be noops. You have to do a bit more to make the subsequent calls wait until the first create call has completed - like using some form of semaphore. Don't know enough Python to suggest details.

if _uri_created[uri]:
  return
_uri_created[uri] = True

Reply:
More specifically:

```
if uri not in self._create_locks:
    # this is first request to create
    self._create_locks[uri] = asyncio.Lock()
    async with self._create_locks[uri]:
        ...this thread does the creating...
        _uri_created[uri] = True
elif not _uri_created[uri]:
    # another thread is in process of creating - wait for that to finish
    while not _uri_created[uri]:
        async with self._create_locks[uri]:
            pass
else:
    # another thread has completed creation - so nothing needs to be done
    pass
```

Reply:
The way I understood this code when reading it is that the lock make sure only one coroutine enters the critical section and then for the next coroutine that can enter the critical section after the first one completed, `_create_for_hash` will be a no-op.

Fully agreed that this is not good code and on the problem with the `del self._create_locks[uri]` you called out above.

We can and need to do better here!

Reply:
cc @architkulkarni @edoakes @AmeerHajAli

Reply:
Indeed something similar to what @sriram-anyscale suggested would be better. Note that this change was a quick/minimal fix for a user bug and the intention is for the code to be restructured to handle the uri caching generically in a higher layer as part of archit's ongoing work (to avoid each plugin needing to roll its own). The PR description has discussion about these points.